### PR TITLE
IDE-4293 Start watching / Stop watching actions should dynamic enable…

### PR DESCRIPTION
**Pull request by jtydhr88**
…d or disabled

### CODE REVIEW

**Comment by yanan** [comment]
it retest

**Comment by yanan** [comment]
ft retest

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 1]
```diff
@@ -47,29 +61,25 @@ public void fillContextMenu(IMenuManager menu) {
...
-		menu.add(new Separator(IWorkbenchActionConstants.MB_ADDITIONS + "-end"));
-	}
 
-	@Override
-	public void init(ICommonActionExtensionSite site) {
-		super.init(site);
+		ICommonActionExtensionSite actionSite = getActionSite();
 
-		ICommonViewerSite viewerSite = site.getViewSite();
+		ICommonViewerSite viewerSite = actionSite.getViewSite();
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]:
```suggestion
		ICommonViewerSite viewSite = actionSite.getViewSite();
```

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 2]
```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
...
+
+		if (selection == null) {
+			return;
+		}
+
 		_watchAction = new WatchWorkspaceModulesAction(provider);
 		_stopAction = new StopWorkspaceModulesAction(provider);
+
+		if (selection instanceof TreeSelection) {
+			TreeSelection watchSelection = (TreeSelection)selection;
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]:
```suggestion
			TreeSelection treeSelection = (TreeSelection)selection;
```

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 3]
```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
...
+
+			List<?> selectProjects = watchSelection.toList();
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]:
```suggestion
			List<?> selectedProjects = watchSelection.toList();
```

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 4]
```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
...
+			}
+
+			Object selectedProject = watchSelection.getFirstElement();
+
+			IWorkspaceProject workspaceProject = LiferayCore.create(IWorkspaceProject.class, selectedProject);
+
+			if (workspaceProject != null) {
+				IProject project = workspaceProject.getProject();
+
+				Set<IProject> watchingProject = workspaceProject.watching();
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]:
```suggestion
				Set<IProject> watchingProjects = workspaceProject.watching();
```

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 5]
```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
...
+
+				String jobName = project.getName() + ":watch";
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]: We need a better way to lookup watch jobs.  @jtydhr88 @Camork we need to do something with qualified names like we did in other tickets.  Can you show @jtydhr88 that technique?

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java [hunk 6]
```diff
@@ -80,9 +90,108 @@ private Separator _invisibleSeparator(String s) {
...
+				).map(
+					selectPath -> {
+						TreePath parentPath = selectPath.getParentPath();
+
+						return parentPath.getLastSegment();
+					}
+				).findFirst(
+				).get();
+
+				IWorkspaceProject parentWorkspaceProject = LiferayCore.create(IWorkspaceProject.class, parentObject);
```
Comment (gamerson) [inline tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]: This logic below looks very much like logic above.  Can this be simplified or extracted?

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java (hunk 5 above)
Comment (Camork) [reply tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]: We can set a specific job family name to the watch tasks in [L145](https://github.com/gamerson/liferay-ide/blob/master/tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleWorkspaceProject.java#L145)

File: tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java (hunk 6 above)
Comment (simonjhy) [reply tools/plugins/com.liferay.ide.gradle.ui/src/com/liferay/ide/gradle/ui/navigator/workspace/LiferayWorkspaceServerActionProvider.java]: I tried to simplified it, but the two part has some different logic. The first part, I can only focus on workspace project, the second, I have to focus on them same time.

**Comment by gamerson** [comment]
sorry @simonjhy  could you rebase again?  thanks

**Comment by jtydhr88** [comment]
@simonjhy please fix the conflicts and resend, thanks.

**Pull request closed without merging**